flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/CodeGenerator.scala

```diff
@@ -1463,21 +1465,23 @@ class CodeGenerator(
     */
   def addReusableFunction(function: UserDefinedFunction): String = {
     val classQualifier = function.getClass.getCanonicalName
-    val fieldTerm = s"function_${classQualifier.replace('.', '$')}"
+    val functionSerializedData = serialize(function)
+    val fieldTerm =
+      s"""
+         |function_${classQualifier.replace('.', '$')}_${DigestUtils.md5Hex(functionSerializedData)}
```

Comment:
I agree with @sunjincheng121.
The naming collision with md5hex are actually intentional to deduplicate identical functions This might happen because the same function is used more than once or because the same function has been registered twice and both are used. By using `CodeGen.newName` we would have the serialized code and initialization as often as the same function is used.